clicommand/artifact_search.go

```diff
@@ -0,0 +1,169 @@
+	Build              string `cli:"build" validate:"required"`
+	IncludeRetriedJobs bool   `cli:"include-retried-jobs"`
+	PrintFormat        string `cli:"printf"`
+
+	// Global flags
+	Debug       bool     `cli:"debug"`
+	NoColor     bool     `cli:"no-color"`
+	Experiments []string `cli:"experiment" normalize:"list"`
+	Profile     string   `cli:"profile"`
+
+	// API config
+	DebugHTTP        bool   `cli:"debug-http"`
+	AgentAccessToken string `cli:"agent-access-token" validate:"required"`
+	Endpoint         string `cli:"endpoint" validate:"required"`
+	NoHTTP2          bool   `cli:"no-http2"`
+}
+
+var ArtifactSearchCommand = cli.Command{
+	Name:        "search",
+	Usage:       "Searches artifacts in Buildkite",
+	Description: SearchHelpDescription,
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "step",
+			Value: "",
+			Usage: "Scope the search to a particular step by using either its name or job ID",
+		},
+		&cli.StringFlag{
+			Name:    "build",
+			Value:   "",
+			EnvVars: []string{"BUILDKITE_BUILD_ID"},
+			Usage:   "The build that the artifacts were uploaded to",
+		},
+		&cli.BoolFlag{
+			Name:    "include-retried-jobs",
+			EnvVars: []string{"BUILDKITE_AGENT_INCLUDE_RETRIED_JOBS"},
+			Usage:   "Include artifacts from retried jobs in the search",
+		},
+		&cli.StringFlag{
+			Name:  "printf",
+			Value: "%j %p %c\n",
+			Usage: `Output formatting of results. Defaults to "%j %p %c\n" (Job ID, path, created at time).
+
+				The following directives are available:
+
+				%i		UUID of the artifact
+
+				%p		Artifact path
+
+				%c		Artifact creation time (an ISO 8601 / RFC-3339 formatted UTC timestamp)
+
+				%j		UUID of the job that uploaded the artifact, helpful for subsequent artifact downloads
+
+				%s		File size of the artifact in bytes
+
+				%S		SHA1 checksum of the artifact
+
+				%u		Download URL for the artifact, though consider using 'buildkite-agent artifact download' instead
+			`,
+		},
+
+		// API Flags
+		AgentAccessTokenFlag,
+		EndpointFlag,
+		NoHTTP2Flag,
+		DebugHTTPFlag,
+
+		// Global flags
+		NoColorFlag,
+		DebugFlag,
+		ExperimentsFlag,
+		ProfileFlag,
+	},
+	Action: func(c *cli.Context) error {
+		// The configuration will be loaded into this struct
+		cfg := ArtifactSearchConfig{}
+
+		l := CreateLogger(&cfg)
+
+		// Load the configuration
+		if err := cliconfig.Load(c, l, &cfg); err != nil {
+			l.Fatal("%s", err)
+		}
+
+		// Setup any global configuration options
+		done := HandleGlobalFlags(l, cfg)
+		defer done()
+
+		// Create the API client
+		client := api.NewClient(l, loadAPIClientConfig(cfg, `AgentAccessToken`))
+
+		// Setup the searcher and try get the artifacts
+		searcher := agent.NewArtifactSearcher(l, client, cfg.Build)
+		artifacts, err := searcher.Search(cfg.Query, cfg.Step, cfg.IncludeRetriedJobs, true)
+		if err != nil {
+			return err
+		}
+
+		artifactCount := len(artifacts)
+
+		if artifactCount == 0 {
+			l.Debug("No matches found for %s", cfg.Query)
```

Comment:
> If the use-case for this command is machine-readability, it might be important to not pollute stdout with debug logging? I believe l.Debug() goes to stdout when debug logging is enabled.

Oh, never mind, as we figured out together the logger goes to stderr:
https://github.com/buildkite/agent/blob/3958a10d1796543d4c4d0cdd565e125706cd0b8d/clicommand/global.go#L93